code/datums/atmosphere/_atmosphere.dm

```diff
@@ -0,0 +1,52 @@
+/datum/atmosphere
+	var/gas_string
+	var/id
+
+	var/list/base_gases // A list of gases to always have
+	var/list/normal_gases // A list of allowed gases:base_amount
+	var/list/restricted_gases // A list of allowed gases like normal_gases but each can only be selected a maximum of one time
+	var/restricted_chance = 10 // Chance per iteration to take from restricted gases
+
+	var/minimum_pressure
+	var/maximum_pressure
+
+	var/minimum_temp
+	var/maximum_temp
+
+/datum/atmosphere/New()
+	generate_gas_string()
+
+/datum/atmosphere/proc/generate_gas_string()
+	var/target_pressure = rand(minimum_pressure, maximum_pressure)
+	var/pressure_scalar = target_pressure / maximum_pressure
+	var/current_pressure = 0
+	var/list/air_types = base_gases?.Copy() || list()
+	for(var/i in air_types)
+		current_pressure += air_types[i]
+	while(current_pressure < target_pressure)
+		var/datum/gas/gastype
+		var/amount
+		if(!prob(restricted_chance))
+			gastype = pick(normal_gases)
+			amount = normal_gases[gastype]
+		else
+			gastype = pick(restricted_gases)
+			amount = restricted_gases[gastype]
+			if(air_types[gastype])
+				continue
+		
+		amount *= rand(50, 200) / 100	// Randomly modifes the amount from half to double the base for some variety
+		amount *= pressure_scalar		// If we pick a really small target pressure we want roughly the same mix but less of it all
+		amount = min(amount, target_pressure-current_pressure)
+		amount = CEILING(amount, 0.1)
+		
+		air_types[gastype] += amount
```

Comment:
no, pv=nrt (gas equation) so the moles you want to add would be **n=pv/rt**

when you add something to the gas mix it just adds moles and temperature, so for example the station has about 104 moles at 293.15 kelvin as defined in the initial mix.

p=(104 moles)(8.31 constant)(293.15 kelvin temperature) / (2500 volume liters) = ~101.3 kpa

https://github.com/tgstation/tgstation/blob/d81ff09046bde0e93cabac1d893b729e693f7d40/code/__DEFINES/atmospherics.dm#L28

https://github.com/tgstation/tgstation/blob/d81ff09046bde0e93cabac1d893b729e693f7d40/code/__DEFINES/atmospherics.dm#L15

just convert this value to moles and then you should be good for at least this part (you'll need to calculate your temperature in a var before this as well to do this just to be clear)

something like
`air_types[gastype] += amount * cell_volume / (r_constant * atmosphere_temperature)`

either way i'd really like to see this in the game but yeah it had mega issues (plus the idea of being able to fuck with random atmospheres sounds cool)